src/AbstractApi.ts

```diff
@@ -1,68 +1,103 @@
 import AuthService from './AuthService';
+import ILoadingProvider from './ILoadingProvider';
+import IInitOptions from './IInitOptions';
 
 export default abstract class AbstractApi {
   readonly apiRoot: string;
   readonly apiRequiresAuth: boolean;
   hasAuthService: boolean = false;
-  private authService?: AuthService;
+  private readonly authService?: AuthService;
+  protected readonly loadingProvider?: ILoadingProvider;
+
+  constructor(apiRoot: string, requiresAuth: boolean, initOptions: IInitOptions = {}) {
+    const {
+      getToken,
+      authScheme,
+      loadingProvider
+    } = initOptions;
 
-  constructor(apiRoot: string, requiresAuth: boolean, getToken?: (...args: any[]) => string, authScheme: AuthSchemes = AuthSchemes.Bearer) {
     this.apiRoot = apiRoot;
     this.apiRequiresAuth = requiresAuth;
     if (requiresAuth) {
       if (!getToken) {
         throw new Error('Must pass token retrieval logic as a function if requiresAuth == true.');
       }
-      this.authService = new AuthService(authScheme, getToken);
+      this.authService = new AuthService(authScheme || AuthSchemes.Bearer, getToken);
       this.hasAuthService = true;
     }
+
+    if (!!loadingProvider) {
+      if (!loadingProvider.onBegin || !loadingProvider.onResolve) {
+        throw new Error('Must provide an `onBegin` and and `onResolve` function to loadingProvider');
+      }
+      this.loadingProvider = loadingProvider;
+    }
   }
 
-  protected abstract async resolve(response: Response): Promise<any>;
+  protected abstract async resolve(response: Response, handleLoading: boolean): Promise<any>;
+
+  async fetch(url: string, options: IRequestOptions): Promise<any> {
+    // we should handle loading by default if there's a loading provider and no `handleLoading`
+    // key in our options object. Otherwise, if there is a loading provider, we go by the value of the
+    // handleLoading option. If no loadingProvider, we will never handle loading.
+    let handleLoading = false;
+    if (!!this.loadingProvider) {
+      handleLoading = !(options.hasOwnProperty('handleLoading') && options.handleLoading === false);
+    }
 
-  async fetch(url: string, options: RequestInit): Promise<any> {
     if (this.apiRequiresAuth) {
       if (!this.authService) {
         throw new Error('Api requires authentication, but AuthService was not initialized.');
       }
 
+      if (handleLoading && !!this.loadingProvider) {
+        this.loadingProvider.onBegin();
+      }
+
       this.authService.setToken();
       options.headers = {
         ...options.headers,
         Authorization: `${this.authService.authScheme} ${this.authService.getToken()}`
       };
     }
 
-
     const root = this.apiRoot.endsWith('/') ? this.apiRoot : this.apiRoot + '/';
     const request = new Request(root + url, options);
-    const response = await fetch(request);
-    return await this.resolve(response);
+    let response;
+    try {
+      response = await fetch(request);
```

Comment:
Why do you mix await and promises here?

Reply:
Promise.reject is just a static method. I'm calling that method rather than throwing because if I were to throw inside a catch block, we wind up with a swallowed error.